### PR TITLE
Replace extractMessage with output from extractApsFrame #1980

### LINKED ISSUE #1980: Replace extractMessage with output from extractApsFrame
#### Problem
`extractMessage` needs to make assumptions about how big the encoding of an `EmberApsFrame` is.  This requires that it be updated whenever encoding/decoding of `EmberApsFrame` changes.

 #### Proposed Solution
It might be better if we just had `extractApsFrame` output how many bytes it consumed so that the consumer can advance the read pointer directly.  0 can mean "error".  Or `extractApsFrame` could output the updated data pointer and remaining length, depending on how we want to structure the callers.

### PR DESCRIPTION
**Pull request by bhaskar-apple**
#### Problem
Replace extractMessage with output from extractApsFrame #1980

 #### Summary of Changes
`extractApsFrame` now returns the number of bytes that was used by the aps header. This lets us remove the hard coded size.

fixes #1980

### CODE REVIEW

File: src/app/chip-zcl-zpro/command-encoder/decoder.c [hunk 1]
```diff
@@ -26,73 +26,73 @@
...
     }
     // Skip first byte, because that's the always-0 frame control.
     uint8_t nextByteToRead = 1;
 
     if (nextByteToRead >= buf_length)
     {
-        return false;
+        return 0;
     }
     memcpy(&outApsFrame->profileId, buffer + nextByteToRead, sizeof(outApsFrame->profileId));
```
Comment (rwalker-apple) [inline src/app/chip-zcl-zpro/command-encoder/decoder.c:43]: is there an issue filed that this code won't work on a BIGENDIAN machine?

Comment (bzbarsky-apple) [reply src/app/chip-zcl-zpro/command-encoder/decoder.c:43]: Yes, https://github.com/project-chip/connectedhomeip/issues/1632

File: examples/chip-tool/main.cpp [hunk 2]
```diff
@@ -100,7 +100,7 @@ static bool ContentMayBeADataModelMessage(System::PacketBuffer * buffer)
 static void HandleDataModelMessage(System::PacketBuffer * buffer)
 {
     EmberApsFrame frame;
-    bool ok = extractApsFrame(buffer->Start(), buffer->DataLength(), &frame);
+    bool ok = extractApsFrame(buffer->Start(), buffer->DataLength(), &frame) > 0 ? true : false;
```
Comment (andy31415) [inline examples/chip-tool/main.cpp:103]: nit: the pattern `foo ? true : false` seems a bit verbose to me and would use `foo` instead. How about:

```
bool ok = (extractApsFrame(buffer->Start(), buffer->DataLength(), &frame) > 0);
```

here and below?

or maybe consde that ok is only used within the if statement and include it as:

```
if (extractApsFrame(buffer->Start(), buffer->DataLength(), &frame) == 0) {
  printf("APS frame processing failure!\n");
  System::PacketBuffer::Free(buffer);
  return;
}

printf("APS frame processing success!\n");
```

although that starts to seem like a larger change, so mabe can be done in a separate PR.

File: src/app/chip-zcl-zpro/command-encoder/decoder.c [hunk 3]
```diff
@@ -103,18 +103,17 @@ void printApsFrame(EmberApsFrame * frame)
...
     uint16_t result = 0;
-    if (msg && buffer_length > kEmberApsFrameSize)
+    EmberApsFrame frame;
+    uint16_t apsFrameSize = extractApsFrame(buffer, buffer_length, &frame);
+    if (msg && apsFrameSize > 0 && buffer_length > apsFrameSize)
     {
         // These are hard coded for now.
-        *msg   = buffer + kEmberApsFrameSize;
-        result = buffer_length - kEmberApsFrameSize;
+        *msg   = buffer + apsFrameSize;
```
Comment (andy31415) [inline src/app/chip-zcl-zpro/command-encoder/decoder.c:115]: should we remove the 'hard coded for now' comment above?

**Review by andy31415 — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@jelderton @saurabhst ?

**Review by saurabhst — APPROVED** [review]

File: src/app/chip-zcl-zpro/command-encoder/chip-zcl-zpro-codec.h [hunk 4]
```diff
@@ -77,8 +77,12 @@ uint16_t encodeReadAttributesCommand(uint8_t * buffer, uint16_t buf_length, uint
 uint16_t encodeReadOnOffCommand(uint8_t * buffer, uint16_t buf_length, uint8_t destination_endpoint);
 
 /** @brief Extracts an aps frame from buffer into outApsFrame
+ * @param buffer Buffer to read from
+ * @param buf_length Length of buffer
+ * @param outApsFrame Pointer to EmberApsFrame struct to read into
+ * @return returns the number of bytes that were consumed to read out the EmberApsFrame. 0 means an error was encountered
  */
-bool extractApsFrame(uint8_t * buffer, uint32_t buf_length, EmberApsFrame * outApsFrame);
+uint16_t extractApsFrame(void * buffer, uint32_t buf_length, EmberApsFrame * outApsFrame);
```
Comment (bzbarsky-apple) [inline src/app/chip-zcl-zpro/command-encoder/chip-zcl-zpro-codec.h:85]: `buffer` arg should stay `uint8_t *`, no?

File: examples/chip-tool/main.cpp [hunk 5]
```diff
@@ -100,18 +100,14 @@ static bool ContentMayBeADataModelMessage(System::PacketBuffer * buffer)
...
+    if (extractApsFrame(buffer->Start(), buffer->DataLength(), &frame) == 0)
     {
         printf("APS frame processing failure!\n");
         System::PacketBuffer::Free(buffer);
         return;
     }
 
+    printf("APS frame processing success!\n");
     uint8_t * message;
     uint16_t messageLen = extractMessage(buffer->Start(), buffer->DataLength(), &message);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/main.cpp:112]: I would have expected this to look more like:
```
uint8_t * message = buffer->Start() + offset;
uint16_t messageLen = buffer->DataLength() - offset;
```
or something, where `offset` is the return value of `extractApsFrame`.  I guess you're leaving `extractMessage` but just reimplementing it on top of `extractApsFrame`?  But if we do that, do we want to change the return type, or have a static non-public `doExtractApsFrame` helper that returns the offset and is called by both `extractApsFrame`, which returns whether the offset is nonzero, and by `extractMessage`?

The tradeoff of doing things the way they're done here is we do all the `extractApsFrame` work twice.  Maybe that's ok?

File: src/app/chip-zcl-zpro/command-encoder/decoder.c [hunk 6]
```diff
@@ -103,18 +103,16 @@ void printApsFrame(EmberApsFrame * frame)
...
-static const size_t kEmberApsFrameSize = 13;
-
 uint16_t extractMessage(uint8_t * buffer, uint16_t buffer_length, uint8_t ** msg)
 {
     // The message starts after the EmberApsFrame.
     uint16_t result = 0;
-    if (msg && buffer_length > kEmberApsFrameSize)
+    EmberApsFrame frame;
+    uint16_t apsFrameSize = extractApsFrame(buffer, buffer_length, &frame);
+    if (msg && apsFrameSize > 0 && buffer_length > apsFrameSize)
```
Comment (bzbarsky-apple) [inline src/app/chip-zcl-zpro/command-encoder/decoder.c:112]: Why do we need the `buffer_length > apsFrameSize` check here?  We _might_ have a length-0 message; presumably trying to process it further will fail is all.  And if `extractApsFrame` is not lying to us, `buffer_length < apsFrameSize` can't happen.

Comment (bhaskar-apple) [reply src/app/chip-zcl-zpro/command-encoder/decoder.c:112]: Yup this check is not needed anymore as this is now inside extractApsFrame

File: examples/chip-tool/main.cpp (hunk 5 above)
Comment (bhaskar-apple) [reply examples/chip-tool/main.cpp:112]: Spoke offline. The cost of using `extractApsFrame` is minimal. So we are keeping it this way for now.

File: src/app/chip-zcl-zpro/command-encoder/chip-zcl-zpro-codec.h (hunk 4 above)
Comment (bhaskar-apple) [reply src/app/chip-zcl-zpro/command-encoder/chip-zcl-zpro-codec.h:85]: Fixing in follow up PR

**Pull request merged**